client/rest/src/main/java/org/elasticsearch/client/DeadHostState.java

```diff
@@ -47,10 +51,19 @@ private DeadHostState() {
      * that failed many consecutive times).
      */
     DeadHostState(DeadHostState previousDeadHostState) {
+        this(previousDeadHostState, TimeSupplier.DEFAULT);
+    }
+
+    DeadHostState(DeadHostState previousDeadHostState, TimeSupplier timeSupplier) {
         long timeoutNanos = (long)Math.min(MIN_CONNECTION_TIMEOUT_NANOS * 2 * Math.pow(2, previousDeadHostState.failedAttempts * 0.5 - 1),
                 MAX_CONNECTION_TIMEOUT_NANOS);
-        this.deadUntilNanos = System.nanoTime() + timeoutNanos;
+        this.deadUntilNanos = timeSupplier.getNanoTime() + timeoutNanos;
         this.failedAttempts = previousDeadHostState.failedAttempts + 1;
+        this.timeSupplier = timeSupplier;
+    }
+
+    boolean shallBeRetried() {
+        return timeSupplier.getNanoTime() - deadUntilNanos > 0;
```

Comment:
Huh! I'd figured you'd pass the `TimeSupplier` in to every method that had to call it. We wouldn't keep references to it around. But that does change the method calls a ton. I think this is a fine way to do it!

Reply:
It is probably worth having javadoc for this though.

Reply:
I wasn't sure initially, but I think this is good as it allows to isolate all the timing related aspects to `DeadHostState`, using the same instance of timesupplier. it feels weird to have to pass in the time supplier here too, or maybe I could do that. Would you prefer to change that?